### ISSUE TITLE
Unable to connect via TLS 1.3

### ISSUE THREAD

**Issue by Tarnum-tst** [openssl/openssl#19051]
Hi!
When building latest commit from master branch it became impossible to connect particular site via TLS 1.3.

I'm using Python 3.7.13 in conjunction with OpenSSL 3.1.0-dev (latest commit),
To reproduce, I'm using this Python script:
```
import socket
import ssl

for j in range (1,10):
  myhost = 'tls13.1d.pw'
  context = ssl.create_default_context()
  with socket.create_connection((myhost, 443)) as mysock:
    with context.wrap_socket(mysock, server_hostname=myhost) as myssock:
      print(str(j) + ' connection to ' + myhost + ' via ' + myssock.version())
```

Got this:
```
Traceback (most recent call last):
  File "mytest.py", line 8, in <module>
    with context.wrap_socket(mysock, server_hostname=myhost) as myssock:
  File "C:\Python37\lib\ssl.py", line 412, in wrap_socket
    session=session
  File "C:\Python37\lib\ssl.py", line 853, in _create
    self.do_handshake()
  File "C:\Python37\lib\ssl.py", line 1117, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: UNABLE_TO_DECODE_ECDH_CERTS] record layer failure (_ssl.c:1091)
```

Everything worked fine
```
1 connection to tls13.1d.pw via TLSv1.3
2 connection to tls13.1d.pw via TLSv1.3
3 connection to tls13.1d.pw via TLSv1.3
4 connection to tls13.1d.pw via TLSv1.3
5 connection to tls13.1d.pw via TLSv1.3
6 connection to tls13.1d.pw via TLSv1.3
7 connection to tls13.1d.pw via TLSv1.3
8 connection to tls13.1d.pw via TLSv1.3
9 connection to tls13.1d.pw via TLSv1.3
```

before this PR: https://github.com/openssl/openssl/pull/18132

**Comment by mattcaswell** [comment]
Hmm. That traceback is virtually useless for finding the real problem. I wonder if python has a way of dumping the whole contents of the OpenSSL error stack - rather than just the last error? The "record layer failure" error just says "something bad happened in the record layer" which isn't very helpful. I wonder if we might need to reconsider this code which is responsible for logging the "record layer failure" error (at least that's one place, and probably the most likely place that this error is coming from).

https://github.com/openssl/openssl/blob/7af110f9f5fb9b039cc09b63768a0b989a7bf5ad/ssl/record/rec_layer_s3.c#L1089-L1096

Perhaps if we don't log an error here python will reveal the next error down in the OpenSSL error stack. Could you try applying the following patch to OpenSSL and then post the traceback you get after that?

```` diff
diff --git a/ssl/record/rec_layer_s3.c b/ssl/record/rec_layer_s3.c
index c187141ee9..d3c721e873 100644
--- a/ssl/record/rec_layer_s3.c
+++ b/ssl/record/rec_layer_s3.c
@@ -1092,7 +1092,7 @@ int ossl_tls_handle_rlayer_return(SSL_CONNECTION *s, int ret, char *file,
             if (al != SSL_AD_NO_ALERT) {
                 ERR_new();
                 ERR_set_debug(file, line, 0);
-                ossl_statem_fatal(s, al, SSL_R_RECORD_LAYER_FAILURE, NULL);
+                ossl_statem_send_fatal(s, al);
             }
             /*
              * else some failure but there is no alert code. We don't log an
````

**Comment by t8m** [comment]
This is reproducible with s_client:

```
LD_LIBRARY_PATH=. apps/openssl s_client -connect tls13.1d.pw:443
Connecting to 194.87.109.56
CONNECTED(00000003)
4007DB90F47F0000:error:0A00010B:SSL routines:tls_validate_record_header:wrong version number:../openssl/ssl/record/methods/tlsany_meth.c:95:
4007DB90F47F0000:error:0A000139:SSL routines::record layer failure:../openssl/ssl/record/rec_layer_s3.c:1235:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 118 bytes and written 730 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
This TLS version forbids renegotiation.
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
```

3.0 works:
```
openssl s_client -connect tls13.1d.pw:443
CONNECTED(00000003)
depth=2 C = US, ST = New Jersey, L = Jersey City, O = The USERTRUST Network, CN = USERTrust ECC Certification Authority
verify return:1
depth=1 C = GB, ST = Greater Manchester, L = Salford, O = Sectigo Limited, CN = Sectigo ECC Domain Validation Secure Server CA
verify return:1
depth=0 CN = tls13.1d.pw
verify return:1
---
Certificate chain
 0 s:CN = tls13.1d.pw
   i:C = GB, ST = Greater Manchester, L = Salford, O = Sectigo Limited, CN = Sectigo ECC Domain Validation Secure Server CA
   a:PKEY: id-ecPublicKey, 384 (bit); sigalg: ecdsa-with-SHA256
   v:NotBefore: Dec 17 00:00:00 2021 GMT; NotAfter: Dec 17 23:59:59 2022 GMT
 1 s:C = GB, ST = Greater Manchester, L = Salford, O = Sectigo Limited, CN = Sectigo ECC Domain Validation Secure Server CA
   i:C = US, ST = New Jersey, L = Jersey City, O = The USERTRUST Network, CN = USERTrust ECC Certification Authority
   a:PKEY: id-ecPublicKey, 256 (bit); sigalg: ecdsa-with-SHA384
   v:NotBefore: Nov  2 00:00:00 2018 GMT; NotAfter: Dec 31 23:59:59 2030 GMT
 2 s:C = US, ST = New Jersey, L = Jersey City, O = The USERTRUST Network, CN = USERTrust ECC Certification Authority
   i:C = GB, ST = Greater Manchester, L = Salford, O = Comodo CA Limited, CN = AAA Certificate Services
   a:PKEY: id-ecPublicKey, 384 (bit); sigalg: RSA-SHA384
   v:NotBefore: Mar 12 00:00:00 2019 GMT; NotAfter: Dec 31 23:59:59 2028 GMT
 3 s:CN = "IP, port: 95.80.230.81:34282"
   i:CN = HS DH: ffdhe3072
   a:PKEY: id-ecPublicKey, 256 (bit); sigalg: ecdsa-with-SHA256
   v:NotBefore: Aug 24 04:25:33 2022 GMT; NotAfter: Sep 23 16:25:33 2022 GMT
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIEwjCCBGigAwIBAgIRAOGq7axHxaGFhKXm0owGBj8wCgYIKoZIzj0EAwIwgY8x
CzAJBgNVBAYTAkdCMRswGQYDVQQIExJHcmVhdGVyIE1hbmNoZXN0ZXIxEDAOBgNV
BAcTB1NhbGZvcmQxGDAWBgNVBAoTD1NlY3RpZ28gTGltaXRlZDE3MDUGA1UEAxMu
U2VjdGlnbyBFQ0MgRG9tYWluIFZhbGlkYXRpb24gU2VjdXJlIFNlcnZlciBDQTAe
Fw0yMTEyMTcwMDAwMDBaFw0yMjEyMTcyMzU5NTlaMBYxFDASBgNVBAMTC3RsczEz
LjFkLnB3MHYwEAYHKoZIzj0CAQYFK4EEACIDYgAE3q3A3tfL+36jumM3GAZTN/nQ
904zMhdlvN+pDecoQoDyGGBhznkmjG2Pqr71M+Kvk8FGDcnALcqHMH2ofdRCLew0
qNzR25+UHVREyaQHXnf7/VFtJoA1zdHnfk1rvxaPo4IC/jCCAvowHwYDVR0jBBgw
FoAU9oUKOxGG4QR9DqoLLNLuzGR7e64wHQYDVR0OBBYEFHFsOH2AShN4DFlKPEoE
2a9o3u8IMA4GA1UdDwEB/wQEAwIHgDAMBgNVHRMBAf8EAjAAMB0GA1UdJQQWMBQG
CCsGAQUFBwMBBggrBgEFBQcDAjBJBgNVHSAEQjBAMDQGCysGAQQBsjEBAgIHMCUw
IwYIKwYBBQUHAgEWF2h0dHBzOi8vc2VjdGlnby5jb20vQ1BTMAgGBmeBDAECATCB
hAYIKwYBBQUHAQEEeDB2ME8GCCsGAQUFBzAChkNodHRwOi8vY3J0LnNlY3RpZ28u
Y29tL1NlY3RpZ29FQ0NEb21haW5WYWxpZGF0aW9uU2VjdXJlU2VydmVyQ0EuY3J0
MCMGCCsGAQUFBzABhhdodHRwOi8vb2NzcC5zZWN0aWdvLmNvbTAnBgNVHREEIDAe
ggt0bHMxMy4xZC5wd4IPd3d3LnRsczEzLjFkLnB3MIIBfgYKKwYBBAHWeQIEAgSC
AW4EggFqAWgAdgBGpVXrdfqRIDC1oolp9PN9ESxBdL79SbiFq/L8cP5tRwAAAX3I
lZKmAAAEAwBHMEUCIQC7Qzg+6KhKp8+oGO5zPmWPNHznYuPcJw0Ef7RAZb4JKQIg
dS86WimZ6BmtJBpmHH+RKNNuLcrKbCqIUDZxA7zeLaQAdwBByMqx3yJGShDGoToJ
QodeTjGLGwPr60vHaPCQYpYG9gAAAX3IlZMWAAAEAwBIMEYCIQCgnYJHb/eTR+NQ
syds56f0HgQ4E+As8vO58uWTLDr1SgIhALNENXDErus+o78sFMRK9vsk8lhnD5IW
g7mJ2L/QP5y0AHUAKXm+8J45OSHwVnOfY6V35b5XfZxgCvj5TV0mXCVdx4QAAAF9
yJWS4wAABAMARjBEAiAMo9CCfR71D7Vt2b7djE9hbzwAnlz6+goXYAIaNr4SHwIg
NRNVuA28ycdu9e/Ss+y1ZBBaMqOOasLSPIUbh/HLcCIwCgYIKoZIzj0EAwIDSAAw
RQIgWIzWrSdeSoyRJw3G7Y+eIpAVMTHqqbjXVg87ubNvwLcCIQDW+aihBLWn3q4i
UVWNzNBPPXJfcIHiFSwfb7x9UsYMhQ==
-----END CERTIFICATE-----
subject=CN = tls13.1d.pw
issuer=C = GB, ST = Greater Manchester, L = Salford, O = Sectigo Limited, CN = Sectigo ECC Domain Validation Secure Server CA
---
No client certificate CA names sent
Peer signing digest: SHA384
Peer signature type: ECDSA
Server Temp Key: DH, 3072 bits
---
SSL handshake has read 4492 bytes and written 1076 bytes
Verification: OK
---
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Server public key is 384 bit
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
```

**Comment by Tarnum-tst** [comment]
@mattcaswell 
"Could you try applying the following patch to OpenSSL and then post the traceback you get after that?"
Done with rebuilding, unfortunately it didn't help much.
```
Traceback (most recent call last):
  File "mytest.py", line 8, in <module>
    with context.wrap_socket(mysock, server_hostname=myhost) as myssock:
  File "C:\Python37\lib\ssl.py", line 412, in wrap_socket
    session=session
  File "C:\Python37\lib\ssl.py", line 853, in _create
    self.do_handshake()
  File "C:\Python37\lib\ssl.py", line 1117, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: A failure in the SSL library occurred (_ssl.c:1091)
```

**Comment by mattcaswell** [comment]
Thanks @t8m. that's really helpful.

Interestingly both the OpenSSL client and the remote server are misbehaving here. Together these things are causing the connection to fail.

The client sends an initial ClientHello and receives an HRR back (which just looks like a ServerHello). The client then retransmits its ClientHello and we get a "real" ServerHello back. Unfortunately the record layer protocol version in the ServerHello is wrong. It *must* be set to 0x0303 but is actually set to 0x0301. The RFC has this to say about it:

>    legacy_record_version:  MUST be set to 0x0303 for all records
      generated by a TLS 1.3 implementation other than an initial
      ClientHello (i.e., one not generated after a HelloRetryRequest),
      where it MAY also be 0x0301 for compatibility purposes.

So, this is definitely incorrect behaviour on the server side. However the RFC goes on to say:

> This field is deprecated and MUST be ignored for all purposes.

So, even though the value is wrong, the OpenSSL client is required to ignore it anyway. It looks like prior to #18132 it did ignore it, but since then it has become stricter.

I need to do more investigation as to why...

**Comment by mattcaswell** [comment]
Fix for this in #19058. Please try it out.

**Comment by Tarnum-tst** [comment]
@mattcaswell 
Just tested, working brilliantly! Thank You very much! :-)
